src/util/expr_util.cpp

```diff
@@ -140,28 +141,35 @@ bool has_subtype(
   const std::function<bool(const typet &)> &pred,
   const namespacet &ns)
 {
-  if(pred(type))
-    return true;
-  else if(type.id() == ID_symbol)
-    return has_subtype(ns.follow(type), pred, ns);
-  else if(type.id() == ID_c_enum_tag)
-    return has_subtype(ns.follow_tag(to_c_enum_tag_type(type)), pred, ns);
-  else if(type.id() == ID_struct || type.id() == ID_union)
-  {
-    const struct_union_typet &struct_union_type = to_struct_union_type(type);
+  std::vector<typet> stack(1, type);
```

Comment:
Start the process with `push_if_not_visited(type);` - otherwise you are going to push (but then not recurse into) the initial type repeatedly if it was a symbol. (Your previous version did an extra, but incorrect, check below.)